payments/providers/bambora_payform.py

```diff
@@ -0,0 +1,242 @@
+        is_valid = True
+        auth_code_calculation_values = [
+            request.GET[param_name]
+            for param_name in ('RETURN_CODE', 'ORDER_NUMBER', 'SETTLED', 'CONTACT_ID', 'INCIDENT_ID')
+            if param_name in request.GET
+        ]
+        correct_auth_code = self.calculate_auth_code('|'.join(auth_code_calculation_values))
+        auth_code = request.GET['AUTHCODE']
+        if not hmac.compare_digest(auth_code, correct_auth_code):
+            logger.warning('Incorrect auth code "{}".'.format(auth_code))
+            is_valid = False
+        return is_valid
+
+    def handle_success_request(self):  # noqa: C901
+        """Handle the payform response after user has completed the payment flow in normal fashion"""
+        request = self.request
+        logger.debug('Handling Bambora user return request, params: {}.'.format(request.GET))
+
+        if not self.check_new_payment_authcode(request):
+            return self.ui_redirect_failure()
+
+        try:
+            order = Order.objects.get(order_number=request.GET['ORDER_NUMBER'])
+        except Order.DoesNotExist:
+            logger.warning('Order does not exist.')
+            return self.ui_redirect_failure()
+
+        return_code = request.GET['RETURN_CODE']
+        if return_code == '0':
+            logger.debug('Payment completed successfully.')
+            try:
+                order.set_state(Order.CONFIRMED, 'Code 0 (payment succeeded) in Bambora Payform success request.')
+                return self.ui_redirect_success(order)
+            except OrderStateTransitionError as oste:
+                logger.warning(oste)
+                order.create_log_entry('Code 0 (payment succeeded) in Bambora Payform success request.')
+                return self.ui_redirect_failure(order)
+        elif return_code == '1':
+            logger.debug('Payment failed.')
+            try:
+                order.set_state(Order.REJECTED, 'Code 1 (payment rejected) in Bambora Payform success request.')
+                return self.ui_redirect_failure(order)
+            except OrderStateTransitionError as oste:
+                logger.warning(oste)
+                order.create_log_entry('Code 1 (payment rejected) in Bambora Payform success request.')
+                return self.ui_redirect_failure(order)
+        elif return_code == '4':
+            logger.debug('Transaction status could not be updated.')
+            order.create_log_entry(
+                'Code 4: Transaction status could not be updated. Use the merchant UI to resolve.'
+            )
+            return self.ui_redirect_failure(order)
+        elif return_code == '10':
+            logger.debug('Maintenance break.')
+            order.create_log_entry('Code 10: Bambora Payform maintenance break')
+            return self.ui_redirect_failure(order)
+        else:
+            logger.warning('Incorrect RETURN_CODE "{}".'.format(return_code))
```

Comment:
Should this be logged as an Order Log Entry too? Could be useful if something funny happens and users are calling and saying "I paid, what happened to my order?" :)

Reply:
👍